spopt/locate/p_median.py

```diff
@@ -0,0 +1,233 @@
+        weights: np.array
+            one-dimensional service load or population demand
+        p_facilities: int
+            number of facilities to be located
+        name: str, default="p-median"
+            name of the problem
+
+        Returns
+        -------
+        PMedian object
+        """
+        r_cli = range(cost_matrix.shape[0])
+        r_fac = range(cost_matrix.shape[1])
+
+        model = pulp.LpProblem(name, pulp.LpMinimize)
+
+        weights = np.reshape(weights, (cost_matrix.shape[0], 1))
+        sij = weights * cost_matrix
+
+        p_median = PMedian(name, model, sij)
+
+        FacilityModelBuilder.add_facility_integer_variable(p_median, r_fac, "y[{i}]")
+        FacilityModelBuilder.add_client_assign_integer_variable(
+            p_median, r_cli, r_fac, "z[{i}_{j}]"
+        )
+
+        p_median.__add_obj(r_cli, r_fac)
+
+        FacilityModelBuilder.add_facility_constraint(
+            p_median, p_median.problem, p_facilities
+        )
+        FacilityModelBuilder.add_assignment_constraint(
+            p_median, p_median.problem, r_fac, r_cli
+        )
+        FacilityModelBuilder.add_opening_constraint(
+            p_median, p_median.problem, r_fac, r_cli
+        )
+
+        return p_median
+
+    @classmethod
+    def from_geodataframe(
+        cls,
+        gdf_demand: GeoDataFrame,
+        gdf_fac: GeoDataFrame,
+        demand_col: str,
+        facility_col: str,
+        weights_cols: str,
+        p_facilities: int,
+        distance_metric: str = "euclidean",
+        name: str = "p-median",
+    ):
+        """
+        Create a PMedian object based on geodataframes. Calculate the cost matrix between demand and facility,
+        and then use from_cost_matrix method.
+
+        Parameters
+        ----------
+        gdf_demand: geopandas.GeoDataFrame
+            demand geodataframe with point geometry
+        gdf_fac: geopandas.GeoDataframe
+            facility geodataframe with point geometry
+        demand_col: str
+            demand geometry column name
+        facility_col: str
+            facility candidate sites geometry column name
+        weights_cols: str
+            weight column name representing service load or demand
+        p_facilities: int
+            number of facilities to be located
+        distance_metric: str, default="euclidean"
+            metrics supported by :method: `scipy.spatial.distance.cdist` used for the distance calculations
+        name: str, default="p-median"
+            name of the problem
+
+        Returns
+        -------
+        PMedian object
+        """
+
+        service_load = gdf_demand[weights_cols].to_numpy()
+        dem = gdf_demand[demand_col]
+        fac = gdf_fac[facility_col]
+
+        dem_type_geom = dem.geom_type.unique()
+        fac_type_geom = fac.geom_type.unique()
+
+        if len(dem_type_geom) > 1 or not "Point" in dem_type_geom:
```

Comment:
We need to be testing for other input geometry types (polygons and lines) since we provide this functionality.

*This also applies to the `PCenter`, `LSCP`, and `MCLP` classes.*